File: Task/Padovan-sequence/C-sharp/padovan-sequence.cs
using System;
using System.Collections.Generic;
using System.Linq;

public static class Padovan
{
    private static readonly List<int> recurrences = new List<int>();
    private static readonly List<int> floors = new List<int>();
    private const double PP = 1.324717957244746025960908854;
    private const double SS = 1.0453567932525329623;

    public static void Main(string[] args)
    {
        for (int i = 0; i < 64; i++)
        {
            recurrences.Add(PadovanRecurrence(i));
            floors.Add(PadovanFloor(i));
        }

        Console.WriteLine("The first 20 terms of the Padovan sequence:");
        recurrences.GetRange(0, 20).ForEach(term => Console.Write(term + " "));
        Console.WriteLine(Environment.NewLine);

        Console.WriteLine("Recurrence and floor functions agree for first 64 terms? " + recurrences.SequenceEqual(floors));
        Console.WriteLine(Environment.NewLine);

        List<string> words = CreateLSystem();

        Console.WriteLine("The first 10 terms of the L-system:");
        words.GetRange(0, 10).ForEach(term => Console.Write(term + " "));
        Console.WriteLine(Environment.NewLine);

        Console.Write("Length of first 32 terms produced from the L-system match Padovan sequence? ");
        List<int> wordLengths = words.Select(s => s.Length).ToList();
        Console.WriteLine(wordLengths.SequenceEqual(recurrences.GetRange(0, 32)));
    }

    private static int PadovanRecurrence(int n)
    {
        if (n <= 2)
            return 1;
        return recurrences[n - 2] + recurrences[n - 3];
    }

    private static int PadovanFloor(int n)
    {
        return (int)Math.Floor(Math.Pow(PP, n - 1) / SS + 0.5);
    }

    private static List<string> CreateLSystem()
    {
        List<string> words = new List<string>();
        string text = "A";

        while (words.Count < 32)
        {
            words.Add(text);
            char[] textChars = text.ToCharArray();
            text = "";
            foreach (char ch in textChars)
            {
                text += ch switch
                {
                    'A' => "B",
                    'B' => "C",
                    'C' => "AB",
                    _ => throw new InvalidOperationException("Unexpected character found: " + ch),
                };
            }
        }

        return words;
    }
}

File: Task/Padovan-sequence/R/padovan-sequence.r
# Function to calculate Padovan sequence iteratively
padovan_sequence <- function(n) {
  recurrences <- numeric(n)
  recurrences[1:3] <- c(1, 1, 1)
  if (n > 3) {
    for (i in 4:n) {
      recurrences[i] <- recurrences[i-2] + recurrences[i-3]
    }
  }
  recurrences
}

# Function to calculate Padovan floor
padovan_floor <- function(aN) {
  PP <- 1.324717957244746025960908854
  SS <- 1.0453567932525329623
  floor((PP^(aN - 1)) / SS + 0.5)
}

# Function to create L-system
create_l_system<- function() {
  words <- c("A")
  text <- "A"

  while (length(words) < 32) {
    text <- strsplit(text, "")[[1]] # Split the string into a list of characters
    text <- sapply(text, function(char) {
      if (char == "A") {
        return("B")
      } else if (char == "B") {
        return("C")
      } else if (char == "C") {
        return("AB")
      }
    })
    text <- paste(text, collapse = "") # Collapse the list back into a single string
    words <- c(words, text) # Append the new word to the list
  }

  words
}


# Main script
num_terms <- 64
padovan_seq <- padovan_sequence(num_terms)
floors <- sapply(0:(num_terms-1), padovan_floor)

cat("The first 20 terms of the Padovan sequence:\n")
cat(padovan_seq[1:20], sep=" ", end="\n")
cat("\n")

cat("Recurrence and floor functions agree for first 64 terms?", all(padovan_seq == floors), "\n\n")

words <- create_l_system()

cat("The first 10 terms of the L-system:\n")
cat(words[1:10], sep=" ", end="\n")
cat("\n")

cat("Length of first 32 terms produced from the L-system match Padovan sequence? ")
word_lengths <- sapply(words, nchar)
cat(all(word_lengths[1:32] == padovan_seq[1:32]))
